sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,550 @@
+        for (unsigned i = 0u; i < chunk; ++i) {
+          dp[i] = isprintable(sp[i]) ? sp[i] : '.';
+        }
+      } else {
+        memcpy(&buf->data[buf->bytes_used], data + idx, chunk);
+      }
+      // Initialize timeout at the point of starting to fill a buffer, since it
+      // will be sent as a complete unit.
+      if (!buf->bytes_used) {
+        s->flush_time = ibex_timeout_init(kUSBLogFlushInterval);
+      }
+      buf->bytes_used += chunk;
+      idx += chunk;
+
+      // Is this packet buffer now ready to be sent?
+      if (buf->bytes_used >= kUSBLogBufferSize) {
+        buffer_completed(ctx, s);
+      }
+    } else if (!s->reliable) {
+      // Blocked awaiting another buffer; drop logging data.
+      break;
+    }
+  } while (idx < len);  // Any more to send?
+
+  return OK_STATUS();
+}
+
+// Interface for known-length (binary) data.
+status_t usb_logging_data(uint8_t s, const uint8_t *data, size_t len) {
+  usb_logging_ctx_t *ctx = &the_ctx;
+  TRY_CHECK(s < ctx->nstreams);
+
+  return usb_logging_send(ctx, &ctx->streams[s], data, len);
+}
+
+// Interface for ASCIIZ text message.
+status_t usb_logging_text(uint8_t s, const char *text) {
+  usb_logging_ctx_t *ctx = &the_ctx;
+  TRY_CHECK(s < ctx->nstreams);
+  TRY_CHECK(text);
+
+  // Note: no strlen() available.
+  uint32_t len = (uint32_t)((char *)memchr(text, 0, SIZE_MAX) - text);
+  return usb_logging_send(ctx, &ctx->streams[s], (const uint8_t *)text, len);
+}
+
+// stdout redirection.
+static size_t usb_log(void *data, const char *buf, size_t len) {
+  usb_logging_stream_t *s = (usb_logging_stream_t *)data;
+  if (s) {
+    usb_logging_ctx_t *ctx = s->ctx;
+    status_t res = usb_logging_send(ctx, s, (uint8_t *)buf, len);
+    if (status_ok(res)) {
+      return len;
+    }
+  }
+  return 0u;
+}
+
+status_t usb_logging_enable(void) {
+  usb_logging_ctx_t *ctx = &the_ctx;
+
+  // Set up a single reliable logging stream; if `ctx->usbutils` is non-zero
+  // then the usb_testutils has already been initialized.
+  // Should be no need to incur the performance hit of remapping characters,
+  // if base_printf/usage is well-behaved.
+  TRY(usb_logging_init(ctx->usbutils, 1u, 1u, 1u, false));
+
+  // Construct the descriptor of our USB sink function for stdout.
+  buffer_sink_t sink;
+  sink.data = &ctx->streams[0];
+  sink.sink = usb_log;
+  // Redirect stdout to us
+  base_set_stdout(sink);
+
+  return OK_STATUS();
+}
+
+status_t usb_logging_init(usb_testutils_ctx_t *usbutils, uint8_t ep_first,
+                          uint8_t nstreams, uint32_t reliable, bool remap) {
+  usb_logging_ctx_t *ctx = &the_ctx;
+
+  // Validate input arguments.
+  TRY_CHECK(nstreams > 0u && nstreams <= kUSBLogMaxStreams);
+
+  // Set up the software layers, if required.
+  if (usbutils) {
+    // usb_testutils layer already initialized, so an endpoint number must be
+    // supplied.
+    ctx->usbutils = usbutils;
+    ctx->inited = false;
```

Comment:
This field is set but not read?

Reply:
Fair enough, thanks. It was used in an earlier version, but I can't reinitialize the USB software and device as intended in usbdev_logging_test because the host processes will lose their connections. Hence, for now at least, the introduction of an explicit 'disconnect' parameter to the _fin() function.